### DISCUSSION TITLE
Laravel Filament Custom Form Field With Livewire Event

### DISCUSSION THREAD

**Discussion by glenelkinsdev** [filamentphp/filament#12692] [Help] [answered]
### Package

Form builder

### Package Version

3

### How can we help you?

I need a custom form field that loads inside a modal during an action on my form, like this:

    Actions\Action::make('Test Action')
        ->label('Test Action')
        ->form([
            FileModal::make('file_id')
    
        ])

Inside the modal my component shows a table which lists files in the database, and each row has a radio input with name file_id, so the user can select one of the files in the table which is then posted back to the main form action:

    ->action(function (array $data, callable $get, callable $set) use ($fileIdFieldName, $textInputName) {
        dd($data);
        
    })

This actually works fine, file_id is present with the id of the file from the selected radio button.

Now i want to implement a livewire search, so i created a custom livewire component that i can put above my table:

    @livewire('attachment-search')

This also works and the search is performed. Now i want the table in the form field to update with the records, so i dispatch an event from my livewire component:

    $this->dispatch('searchResults',['results'=>FileRepository::make()->getAll(limit: 10,searchTerm: trim($this->searchTerm))]);

I am then using $listeners in the form component:

    protected array $listeners = ['searchResults'=>'filterResults'];

This is where the problem is, the filterResults() never fires, as in it's not picking up the event.

I'm obviously doing this wrong but i'm not sure on the approach to make it work. I did try moving the table directly inside the livewire search component to update the table, which worked, but that then caused the form in the modal to post a null on file_id , so it's clear i can't have the form inside the livewire component to post back to the form outside of the modal

So in summary:

1) I have a form inside a resource page that includes a custom form field
2) My custom field has a function getRecords() which loads records from db and displays them in a table in the action modal
3) I have a livewire component @livewire('attachment-search') which simply has an input field: <input type="text" wire:model.live="searchTerm"/>
4) My livewire component reacts to the field change and dispatches an event "searchResults" to pass the results back to the modal form field component
5) My form field component has protected array $listeners = ['searchResults'=>'searchResults'];
6) **The problem: My form field component does not react to the event so i can't update the table with the search results**

I'm obviously doing this wrong, if i set the listener in the resource itself it works, i added the method to a trait, but this isn't an ideal way of doing it, i need the modal itself to listen to the event so it can be easily used anywhere

**Comment by leandrocfe** [comment, marked as answer]
1 - The form field isn't a Livewire component; the pages (list, create, edit, etc.) are Livewire components. You should register the event within the Livewire component.
2 - If you're using Livewire 3, you should follow this approach: https://livewire.laravel.com/docs/events

**Reply by glenelkinsdev** [reply]
I did think this, which is why it works when i add the listener to the resource component, but this really isn't an ideal solution as i will need to add the trait and listener to every resource that uses my modal component

**Reply by leandrocfe** [reply]
You can also register the listener in the field.. something like this:

```php
YourField::make('xx')
    ->registerListeners(['searchResults' => [function (YourField $component) {
         //...
    }]])
```

**Reply by glenelkinsdev** [reply]
Still not ideal, it would be nicer if the field itself could simply define the listener as everywhere the component is being used performs the exact same action. It actually works nicer if i use my trait and include that in every resource that uses it. But i would prefer it if there was a way for the component itself to handle it - if it's not possible it's not possible, so i've done this...

```
<?php

namespace App\Traits;

/**
 *
 */
trait AttachmentSearchTrait
{

    /**
     * @var array
     */
    public array $attachments = [];

    /**
     * @param array $result
     * @return void
     */
    public function attachmentsSearchResults(array $result) : void
    {
        $this->attachments = $result['results'];

    }
}
```

```
class CreateIngredient extends CreateRecord
{

    use AttachmentSearchTrait;

    /**
     * @var string
     */
    protected static string $resource = IngredientResource::class;

    /**
     * @var string[]
     */
    protected $listeners = ['attachmentsSearchResults'=>'attachmentsSearchResults'];

}
```

This gets it working on any resource i need it to, but still not an ideal solution

**Reply by leandrocfe** [reply]
Did you try `registerListeners` ?

**Reply by glenelkinsdev** [reply]
I know that works, but i don't want to repeatedly register the listener each time i use the field, if that makes sense?

**Reply by borna453** [reply]
@glenelkinsdev late reply, but i think in the AppServiceProvider, you can customize the field for each usage.
Example : 
`        DatePicker::configureUsing(function (DatePicker $datePicker) {
            $datePicker
                ->native(false)
                ->displayFormat('d-m-Y')
                ->closeOnDateSelection()
                ->placeholder('dd-mm-yyyy');
        });`
        
        This is just my example for configuring the DatePicker

**Comment by dmitry-udod** [comment]
Hi! Any solution for this for Filament V4, seem `registerListeners` for form component was removed ? :(

**Reply by allmanaj** [reply]
I dug into the code and poster my workaround [here](https://github.com/filamentphp/filament/discussions/18415#discussioncomment-15477555)
